### PR TITLE
CDAP-8355 Support disabling explore queries for impersonated namespace

### PR DESCRIPTION
**Pull request by rohitsinha54**
- Adds a config ``explore.as.principal`` in NamespaceConfig which is used to determine whether the explore queries can run in the namespace impersonating the namespace owner. 
- By default value is true representing that explore queries can run by impersonating the namespace owner.
- User can specify it to be false to disallow running explore queries in the namespace by impersonating namespace owner.
- Support for this in CLI.

Note: This is not yet tested on cluster. I am doing it now.

Issue: https://issues.cask.co/browse/CDAP-8355
Build: http://builds.cask.co/browse/CDAP-RUT530-1

### CODE REVIEW

File: cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java [hunk 1]
```diff
@@ -151,7 +151,49 @@ public void testNamespaces() throws Exception {
...
+    // now check with just key tab uri
+    namespaceMeta = new NamespaceMeta.Builder().setName("test_ns").setKeytabURI("/some/path").build();
+    try {
+      namespaceAdmin.create(namespaceMeta);
+      Assert.fail();
+    } catch (BadRequestException bre) {
+      // expected
+    }
+
+    // if set explore as principal is set to false it shoule be present with both keytab uri and principal
```
Comment (anew) [inline cdap-app-fabric/src/test/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdminTest.java]: shoule/should

File: cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java [hunk 2]
```diff
@@ -45,6 +46,9 @@
   @SerializedName(HIVE_DATABASE)
   private final String hiveDatabase;
 
+  @SerializedName(EXPLORE_AS_PRINCIPAL)
+  private final boolean exploreAsPrincipal;
```
Comment (anew) [inline cdap-proto/src/main/java/co/cask/cdap/proto/NamespaceConfig.java]: How do you ensure that this defaults to true if it is created from Json?

**Review by anew — COMMENTED** [review]
Only two comments, rest looks good.

**Comment by anew** [comment]
Looks like you also need to change NamespaceCommandUtils.prettyPrintNamespaceConfigCLI() to include the new field. I actually forgot to do that for groupName, can you add that as well

**Comment by rohitsinha54** [comment]
Resolved conflicts. Restarted build.

**Comment by anew** [comment]
LGTM

**Comment by rohitsinha54** [comment]
While testing this on cluster I noticed failure to run explore queries on stream in a custom mapped namespace. I am looking into that as of now and the PR will merged after we fix this new issue and complete the test.

File: cdap-common/src/main/java/co/cask/cdap/common/kerberos/OwnerAdmin.java [hunk 3]
```diff
@@ -106,6 +106,30 @@ void add(NamespacedEntityId entityId, KerberosPrincipalId kerberosPrincipalId)
   ImpersonationInfo getImpersonationInfo(NamespacedEntityId entityId) throws IOException;
```
Comment (anew) [inline cdap-common/src/main/java/co/cask/cdap/common/kerberos/OwnerAdmin.java]: Looks like this one is unused now. You can remove it.

File: cdap-security/src/main/java/co/cask/cdap/security/impersonation/Impersonator.java [hunk 4]
```diff
@@ -46,6 +47,21 @@
...
+   * @param entityId the entity to use to lookup the user to impersonate
+   * @param callable the callable to execute
+   * @param <T> return type of the callable
+   * @param impersonatedOpType {@link ImpersonatedOpType} representing the type of the operation which is
+   * being impersonated
+   *
+   * @return the return value of the callable
+   * @throws Exception if the callable throws any exception
+   */
+  <T> T doAs(final NamespacedEntityId entityId, final Callable<T> callable,
```
Comment (anew) [inline cdap-security/src/main/java/co/cask/cdap/security/impersonation/Impersonator.java]: remove the final from interface?

**Review by anew — COMMENTED** [review]
Two more comments, plus this:
- can you fix the NamespaceHttpHandler to validate the JSON for the PUT to .../properties?
- can you log that the namespace was updated, and what was updated?

**Comment by rohitsinha54** [comment]
BUILD: http://builds.cask.co/browse/CDAP-RUT530-8

File: cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java [hunk 5]
```diff
@@ -158,4 +153,14 @@ private boolean isReserved(String namespaceId) {
     return Id.Namespace.DEFAULT.getId().equals(namespaceId) || Id.Namespace.SYSTEM.getId().equals(namespaceId) ||
       Id.Namespace.CDAP.getId().equals(namespaceId);
   }
+
+  private NamespaceMeta getNamespaceMeta(HttpRequest request) throws BadRequestException {
+    NamespaceMeta metadata;
+    try {
+      metadata = parseBody(request, NamespaceMeta.class);
```
Comment (anew) [inline cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/NamespaceHttpHandler.java]: simpler: return parseBody(...)

**Review by anew — COMMENTED** [review]
one final comment, then LGTM

**Comment by rohitsinha54** [comment]
Build passed merging.

**Pull request merged**